File: project_euler_unified/problem_709/solution.py
"""
Problem 709: Even Stevens
"""

def solve():
    MOD = 10**9 + 7
    N = 10**7
    inv2 = pow(2, MOD - 2, MOD)
    inv3 = pow(3, MOD - 2, MOD)

    # p(n) = 2/3 + (1/3)(-1/2)^n
    # sum_{n=1}^{N} p(n) = 2N/3 + (1/3) * sum_{n=1}^{N} (-1/2)^n
    # Geometric sum: sum = (-1/2)(1 - (-1/2)^N) / (1 - (-1/2)) = (-1/2)(1-(-1/2)^N)/(3/2)
    # = -(1-(-1/2)^N)/3

    two_N_over_3 = 2 * N % MOD * inv3 % MOD

    # (-1/2)^N mod MOD
    neg_half = MOD - inv2  # -1/2 mod MOD
    neg_half_N = pow(neg_half, N, MOD)

    geo_sum = (MOD - 1) * (1 - neg_half_N + MOD) % MOD * inv3 % MOD

    result = (two_N_over_3 + inv3 * geo_sum) % MOD
    return result

answer = solve()
print(answer)

File: project_euler_unified/problem_709/solution.cpp
#include <bits/stdc++.h>
using namespace std;
const long long MOD = 1e9 + 7;
long long power(long long a, long long b, long long mod) {
    long long res = 1; a %= mod;
    while (b > 0) {
        if (b & 1) res = res * a % mod;
        a = a * a % mod;
        b >>= 1;
    }
    return res;
}
int main() {
    long long N = 10000000;
    long long inv2 = power(2, MOD - 2, MOD);
    long long inv3 = power(3, MOD - 2, MOD);
    long long two_N_3 = 2 * N % MOD * inv3 % MOD;
    long long neg_half = (MOD - inv2) % MOD;
    long long neg_half_N = power(neg_half, N, MOD);
    long long geo = (MOD - 1 + MOD) % MOD * ((1 - neg_half_N + MOD) % MOD) % MOD * inv3 % MOD;
    long long ans = (two_N_3 + inv3 % MOD * geo % MOD) % MOD;
    cout << ans << endl;
    return 0;
}
